src/checkmatelib/url/canonicalize.py

```diff
@@ -24,12 +24,26 @@ def canonicalize(cls, url):
         :param url: URL to normalise
         :return: A normalised version of the URL
         """
-        parts = cls.canonical_split(url)
-        clean_url = urlunparse(parts)
+        return cls.canonical_join(cls.canonical_split(url))
+
+    @classmethod
+    def canonical_join(cls, parts):
+        """Join canonical parts into a URL.
+
+        This assumes you have called `canonical_split` to get the parts, and
+        does not perform any cleaning itself. The fragment is always ignored,
+        but is present to support the same interface as `urllib`
+
+        :param parts: Tuple of (scheme, netloc, path, params, query, fragment)
+        :return: A single URL string
+        """
+        scheme, netloc, path, params, query, _fragment = parts
+
+        clean_url = urlunparse([scheme, netloc, path, params, query, None])
 
         # Get around the fact that URL parse strips off the '?' if the query
         # string is empty
-        if url.endswith("?") and not clean_url.endswith("?"):
+        if not query and query is not None:
             clean_url += "?"
```

Comment:
We now set query to `""` for empty and `None` for absent in split.

Reply:
Does adding this present-but-empty query have any affect downstream? Is it possible that Checkmate would block a URL with an empty query but allow the same URL with an absent query for example?

Reply:
I'm not sure honestly. It's conceivable, but unlikely I think.

This was originally intended to match Google's URL filtering data set, which (like ours now because we copied it) works on hashes of the URL. These hashes are taken at multiple levels.

One of those levels will include the `?`, so this could only make a difference in a pretty wacky scenario where:

 * The URL is like this in the first place
 * The URL was specified to this level in the rules
 * Google does this too (the cannibalisation method is a bit under-specified)

Most rules apply at the domain or domain + short amount of path level.

Reply:
This is actually explicitly covered in the examples provided by Google:

 * https://cloud.google.com/web-risk/docs/urls-hashing#to_canonicalize_the_path
 
`Canonicalize("http://www.google.com/q?") = "http://www.google.com/q?";`

So it is part of the spec